File: 0017-letter-combinations-of-a-phone-number/0017-letter-combinations-of-a-phone-number.java
class Solution {
    public List<String> letterCombinations(String digits) {
        
        ArrayList<String> names = new ArrayList<>();
        if(digits.isEmpty()){
            return names;
        }
        pad("",digits,names);
        return names;
    }
    private static void pad(String p,String up,ArrayList<String> list){

        String[] map ={
            "","","abc","def","ghi","jkl","mno","pqrs","tuv","wxyz"
        };
        if(up.isEmpty()){
            System.out.println(p);
            list.add(p);
            return;
        }

        int digit = up.charAt(0)-'0';
        String letters = map[digit];
        for(int i=0;i<letters.length();i++){
            char ch = letters.charAt(i);
            pad(p+ch,up.substring(1),list);
        }
         
    }
}

File: 0017-letter-combinations-of-a-phone-number/0017-letter-combinations-of-a-phone-number.cpp
class Solution {
public:
    vector<string> temp;
    unordered_map<char, string> mp = {
        {'2', "abc"},
        {'3', "def"},
        {'4', "ghi"},
        {'5', "jkl"},
        {'6', "mno"},
        {'7', "pqrs"},
        {'8', "tuv"},
        {'9', "wxyz"}
    };
    vector<string> letterCombinations(string digits) {
        generate(digits,"",0);
        return temp;
    }

    void generate(string &digits,string curr,int index){
        if(index==digits.size()){
            temp.push_back(curr);
            return;
        }

        string let=mp[digits[index]];

        for(char c:let){
            generate(digits,curr+c,index+1);
        }
    }
};